Makefile

```diff
@@ -12,130 +12,220 @@
+# BINS is the set of built command executables.
+BINS = aws-k8s-agent aws-cni grpc-health-probe cni-metrics-helper
+# DOCKER_ARGS is extra arguments passed during container image build.
+DOCKER_ARGS =
+# DOCKER_RUN_FLAGS is set the flags passed during runs of containers.
+DOCKER_RUN_FLAGS = --rm -ti $(DOCKER_ARGS)
+# DOCKER_BUILD_FLAGS is the set of flags passed during container image builds
+# based on the requested build.
+DOCKER_BUILD_FLAGS = --build-arg GOARCH="$(ARCH)" \
+					  --build-arg docker_arch="$(DOCKER_ARCH)" \
+					  --build-arg golang_image="$(GOLANG_IMAGE)" \
+					  --network=host \
+	  		          $(DOCKER_ARGS)
+
+# Default to building an executable using the host's Go toolchain.
+.DEFAULT_GOAL = build-linux
+
+# Build both CNI and metrics helper container images.
+all: docker docker-metrics
 
-ifeq ($(ARCH),aarch64)
-  ARCH = arm64
-else
-endif
-ifeq ($(ARCH),x86_64)
-  ARCH = amd64
-endif
+dist: all
+	mkdir -p $(DESTDIR)
+	docker save $(IMAGE_NAME) | gzip > $(IMAGE_DIST)
+	docker save $(METRICS_IMAGE_NAME) | gzip > $(METRICS_IMAGE_DIST)
 
-# Default to build the Linux binary
+# Build the VPC CNI plugin agent using the host's Go toolchain.
+build-linux: BUILD_FLAGS = -mod=readonly -ldflags '-s -w $(LDFLAGS)'
 build-linux:
-	GOOS=linux GOARCH=$(ARCH) CGO_ENABLED=0 go build -o aws-k8s-agent -ldflags "-s -w $(LDFLAGS)" ./cmd/aws-k8s-agent/
-	GOOS=linux GOARCH=$(ARCH) CGO_ENABLED=0 go build -o aws-cni -ldflags " -s -w $(LDFLAGS)" ./cmd/routed-eni-cni-plugin/
-	GOOS=linux GOARCH=$(ARCH) CGO_ENABLED=0 go build -o grpc-health-probe -ldflags "-s -w $(LDFLAGS)" ./cmd/grpc-health-probe/
-
-# Download portmap plugin
-download-portmap:
-	mkdir -p tmp/downloads
-	mkdir -p tmp/plugins
-	curl -L -o tmp/downloads/cni-plugins-$(ARCH).tgz https://github.com/containernetworking/plugins/releases/download/v0.7.5/cni-plugins-$(ARCH)-v0.7.5.tgz
-	tar -vxf tmp/downloads/cni-plugins-$(ARCH).tgz -C tmp/plugins
-	cp tmp/plugins/portmap .
-	rm -rf tmp
-
-# Build CNI Docker image
+	go build $(BUILD_FLAGS) -o aws-k8s-agent     ./cmd/aws-k8s-agent
+	go build $(BUILD_FLAGS) -o aws-cni           ./cmd/routed-eni-cni-plugin
+	go build $(BUILD_FLAGS) -o grpc-health-probe ./cmd/grpc-health-probe
+
+# Build VPC CNI plugin & agent container image.
 docker:
-	@docker build $(DOCKER_ARGS) --build-arg arch="$(ARCH)" -f scripts/dockerfiles/Dockerfile.release -t "$(IMAGE):$(VERSION)" .
-	@echo "Built Docker image \"$(IMAGE):$(VERSION)\""
+	docker build $(DOCKER_BUILD_FLAGS) \
+		-f scripts/dockerfiles/Dockerfile.release \
+		-t "$(IMAGE_NAME)" \
+		.
+	@echo "Built Docker image \"$(IMAGE_NAME)\""
 
+# Run the built cni container image to use in functional testing
 docker-func-test: docker
-	docker run $(DOCKER_ARGS) -it "$(IMAGE):$(VERSION)"
+	docker run $(DOCKER_RUN_FLAGS) \
+		"$(IMAGE_NAME)"
 
-# unit-test
+# Run unit tests
 unit-test:
-	GOOS=linux CGO_ENABLED=1 go test -v -cover $(ALLPKGS)
+	go test -mod=readonly -v -cover $(ALLPKGS)
```

Comment:
We will need to run `go mod tidy` before using `-mod=readonly`...

https://www.reddit.com/r/golang/comments/c2ckjz/why_does_my_gosum_file_keep_changing/